### PR TITLE
Flow implicit size

### PR DESCRIPTION
**Pull request by stephenmdangelo**
Note, the check for `this.flow === undefined` is required because `Positioner` calls `Flow.layoutChildren` in its constructor, but the `flow` property and `Flow` enum aren't registered until later.  

`Flow` was already calling `layoutChildren` at the end of it's constructor, so this doesn't break anything, but allows us to use the `Flow` enum within `layoutChildren` (instead of the check `this.flow == 0`).

I don't know if there's a better way to protect against this in general.  Unless enums/properties are created BEFORE the call to `meta.super`, but I don't know if that breaks things in the general case.

In the analogous example in C++ where `layoutChildren` would be a pure virtual function,  the compiler warns you and you get a linking error: http://cpp.sh/5jrm. Not sure if this just means `Positioner` should just leave the responsibility of calling `layoutChildren` to it's child classes, or if this is a larger issue that can be solved generically for other classes as well?

### CODE REVIEW

**Comment by stephenmdangelo** [comment]
Wait on this, I think the implicitHeight/Width may be incorrect if the opposite value is explicitly set.

**Comment by stephenmdangelo** [comment]
Fixed now, added the two cases that weren't working to the test.

**Comment by ChALkeR** [comment]
Does this affect only `Flow`? I remember seeing other similar code in the repo.

**Comment by stephenmdangelo** [comment]
> Does this affect only Flow? I remember seeing other similar code in the repo (perhaps Grid?).

What bit are you referring to, the bit about calling `layoutChildren` in the `Positioner` constructor? `Grid.layoutChildren` would have the same problem, but avoids it by comparing to `0`, `1` instead of referencing `Grid.LeftToRight` or `Grid.TopToBottom`.

I believe the `implicitWidth` and `implicitHeight` of `Grid` are correct since it has explicit rows/columns, not the implicit ones in `Flow` based on it's width/height.

File: src/modules/QtQuick/Flow.js [hunk 1]
```diff
@@ -55,9 +61,12 @@ registerQmlType({
             curVPos += child.height + this.spacing;
         }
     }
-    if (this.flow == 0)
-        this.implicitHeight = curVPos + rowSize;
-    else
-        this.implicitWidth = curHPos + rowSize;
+    if (this.flow === Flow.LeftToRight) {
```
Comment (ChALkeR) [inline src/modules/QtQuick/Flow.js]: `this.Flow.LeftToRight`, let's not rely on globals.

File: src/modules/QtQuick/Flow.js [hunk 2]
```diff
@@ -17,10 +17,16 @@ registerQmlType({
     this.flowChanged.connect(this, this.layoutChildren);
     this.layoutDirectionChanged.connect(this, this.layoutChildren);
     this.widthChanged.connect(this, this.layoutChildren);
+    this.heightChanged.connect(this, this.layoutChildren);
     this.layoutChildren();
   }
 
   layoutChildren() {
+    if (this.flow === undefined) {
+      /* Flow has not been fully initialized yet */
```
Comment (ChALkeR) [inline src/modules/QtQuick/Flow.js]: Could you use `//` for single-line comments?

**Comment by ChALkeR** [comment]
LGTM. This needs to be rebased, though.

File: src/modules/QtQuick/Flow.js (hunk 1 above)
Comment (ChALkeR) [inline src/modules/QtQuick/Flow.js]: See https://github.com/qmlweb/qmlweb/commit/032f73242906e6a19edff1189367ebbce52e6d30, for example.

**Comment by stephenmdangelo** [comment]
Rebase, and changes applied.

**Comment by ChALkeR** [comment]
Merged in f57d67a...6d9b2314, thanks!

**Pull request closed without merging**